lib/spack/spack/package.py

```diff
@@ -1274,6 +1285,10 @@ def do_install(self,
                     **kwargs
                 )
 
+        if self.try_install_from_binary_cache(explicit):
```

Comment:
Does installation throw an exception if any of download, verification, or database modification fails? Should that be caught here or inside of `try_install_from_binary_cache`?

Reply:
I am planning on making it optional to auto-retrieve packages from a binary cache (I just added a TODO to the PR description). I think it will be sufficient to allow the user to disable it altogether instead of trying to recover from a failed download or verification. While I think another PR could try to recover from download/verification failures, I don't think that failed DB modifications should be recoverable.